src/main/java/org/ow2/proactive/service/GetFromURLService.java

```diff
@@ -0,0 +1,90 @@
+ * the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.service;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+
+
+@Service
+public class GetFromURLService {
+
+    @Autowired
+    private CatalogObjectService catalogObjectService;
+
+    @Autowired
+    private RemoteObjectService remoteObjectService;
+
+    @Autowired
+    private ServiceConfiguration serviceConfiguration;
+
+    private static final String GET_FROM_URL = "PA:GET_FROM_URL";
+    private static final String URL_REGEX = "(.*?)";
+    private static final String CATCH_URL_REGEX_WITH_HTML_QUOTE = "&quot;"+ URL_REGEX + "&quot;";
+    private static final String CATCH_URL_REGEX = "\\\""+ URL_REGEX + "\\\"";
+    private static final String GET_FROM_URL_PATTERN = GET_FROM_URL+"\\((("+CATCH_URL_REGEX+")|("+CATCH_URL_REGEX_WITH_HTML_QUOTE+"))\\)";
+
+    /**
+     * Get a resource from the catalog and resolve PA:GET_FROM_URL("url") if necessary
+     * @param bucketId is the resource bucket id
+     * @param myResourceID is the resource name
+     * @param resolveLinks on true replace PA:GET_FROM_URL("url") by its value otherwise return the raw resource
+     * @return a string which contains the
+     */
+    public String getResource(long bucketId, String myResourceID, boolean resolveLinks) {
+
+        Pattern pattern = Pattern.compile(GET_FROM_URL_PATTERN);
+
+        String resource = catalogObjectService.getRawCatalogObject(serviceConfiguration.getCatalogURL(),bucketId,myResourceID);
+        if(! resolveLinks){
+            return resource;
+        }
+        Matcher tokenMatcher = pattern.matcher(resource);
+        String resourceURL;
+        while(tokenMatcher.find()){
+            resourceURL = extractURLFromToken(tokenMatcher.group());
+            resource = tokenMatcher.replaceFirst((String) remoteObjectService.sendRequest(resourceURL,String.class));
+            tokenMatcher.reset(resource);
+        }
+        return resource;
+    }
+
+    private String extractURLFromToken(String token){
+        if(token.endsWith("\")")){
+            token = token.replaceFirst(GET_FROM_URL+"\\(\"","");
+            token = token.replaceFirst("\"\\)","");
+        }else{
+            token = token.replaceFirst(GET_FROM_URL+"\\(&quot;","");
+            token = token.replaceFirst("&quot;\\)","");
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Introduce a new variable instead of reusing the parameter "token". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1226)